### PR TITLE
patch redis cluster multi-key operations

### PR DESCRIPTION
**Pull request by joeyorlando**
# Which issue(s) this PR fixes

Related to https://github.com/grafana/oncall-private/issues/2363

Addresses this issue that arises when using `cache.get_many`/`cache.set_many` operations with a Redis Cluster:
```python3
File "/usr/local/lib/python3.11/site-packages/redis/cluster.py", line 1006, in determine_slot
    raise RedisClusterException(
redis.exceptions.RedisClusterException: MGET - all keys must map to the same key slot
```

From the Redis Cluster [docs](https://redis.io/docs/reference/cluster-spec/#hash-tags), this can be addressed with this 👇 . Basically this will ensure that keys in multi-key operations will resolve to the same hash slot (read: node):

> Hash tags
> There is an exception for the computation of the hash slot that is used in order to implement hash tags. Hash tags are a way to ensure that multiple keys are allocated in the same hash slot. This is used in order to implement multi-key operations in Redis Cluster.
> 
> To implement hash tags, the hash slot for a key is computed in a slightly different way in certain conditions. If the key contains a "{...}" pattern only the substring between { and } is hashed in order to obtain the hash slot. However since it is possible that there are multiple occurrences of { or } the algorithm is well specified by the following rules:
> 
> IF the key contains a { character.
> AND IF there is a } character to the right of {.
> AND IF there are one or more characters between the first occurrence of { and the first occurrence of }.
> Then instead of hashing the key, only what is between the first occurrence of { and the following first occurrence of } is hashed.

## Checklist

- [x] Unit, integration, and e2e (if applicable) tests updated
- [ ] Documentation added (or `pr:no public docs` PR label added if not required)
- [ ] `CHANGELOG.md` updated (or `pr:no changelog` PR label added if not required)

### CODE REVIEW

File: engine/apps/metrics_exporter/helpers.py
Comment (joeyorlando) [inline engine/apps/metrics_exporter/helpers.py:1]: these were the only metrics related keys that were used in `cache.get_many` operations

**Review by matiasb — APPROVED** [review]
LGTM, with a couple of questions

File: engine/apps/schedules/ical_utils.py [hunk 1]
```diff
@@ -403,15 +404,24 @@ def get_cached_oncall_users_for_multiple_schedules(schedules: typing.List["OnCal
...
+            f"{CACHE_KEY_PREFIX}{schedule.public_primary_key}", CACHE_KEY_PREFIX
+        )
 
     def _get_schedule_public_primary_key_from_schedule_oncall_users_cache_key(cache_key: str) -> str:
-        return cache_key.replace("schedule_", "").replace("_oncall_users", "")
+        """
+        remove any brackets that might be included in the cache key (when redis cluster is active).
+        See `_generate_cache_key_for_schedule_oncall_users` just above
+        """
+        cache_key = cache_key.replace("{", "").replace("}", "")
```
[the remark this answers is not in the document]
Comment (joeyorlando) [reply engine/apps/schedules/ical_utils.py:419]: that's a great question, I _assume_ so? (fwiw I've never seen a schedule public primary key that had non-alphanumeric characters)

**Comment by joeyorlando** [comment]
> fwiw, we may want to consider using django-redis [bulk scan](https://github.com/jazzband/django-redis#scan--delete-keys-in-bulk)

@matiasb I _think_ from what I understand in the linked docs section, `keys` and `iter_keys` refer to a [`KEYS` `redis` command](https://redis.io/commands/keys/) (which afaik will only return cache _keys_ that match the glob pattern, not values).

I believe `get_many` and `set_many` map to the [`MGET`](https://redis.io/commands/mget/) and [`MSET`](https://redis.io/commands/mset/) commands, respectively.

**Pull request merged**